### ISSUE TITLE
[bug] Cross compiling to bare metal targets does not work with CMakeToolchain

### ISSUE THREAD

**Issue by serghov** [conan-io/conan#11041]
Conan has os == 'Arduino' type for bare metal compilation, however when generating a project with CMakeToolchain generator and a bare metal compiler cmake fails on compiler check stage.

```
-- Check for working CXX compiler: /usr/local/bin/arm-none-eabi-g++
-- broken
```

Here is the profile I am trying to write for bare metal compilation

```
include(default)
[settings]
arch=armv6
os=Arduino
os.board=none
compiler=gcc
compiler.version=11.2
compiler.libcxx=libstdc++11
build_type=Release
[conf]
tools.build:cxxflags=[]
tools.build:cflags+=[]
[env]
CC=/usr/local/bin/arm-none-eabi-gcc
CXX=/usr/local/bin/arm-none-eabi-g++
CMAKE_SYSTEM_PROCESSOR=arm
```

The solution would be for conan to add `set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)` to conan_toolchain.cmake while generating.

The whole 'Arduino' system as far as I understand is very underdeveloped in conan, however this fix would make cross compiling for embedded targets much easier.

**Comment by czoido** [comment]
Hi @serghov,
Thanks for the suggestion, we will check if that case is worth adding to the CMakeToolchain generator.
Anyway, you could use the `tools.cmake.cmaketoolchain:toolchain_file=<filepath>` configuration to use your own toolchain and set that variable in CMake. Please, [check the Conan documentation](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html#using-a-custom-toolchain-file) for more detailed information.

**Comment by memsharded** [comment]
I think that for Conan 2.0, ``Arduino`` shouldn't be a first level "os", but a subsetting of "baremetal"

**Comment by dkoerner-festo** [comment]
> I think that for Conan 2.0, `Arduino` shouldn't be a first level "os", but a subsetting of "baremetal"

@memsharded Why? We use RTOSs like ThreadX and my first approach would be to define it as "os" and not as a second level of "baremetal". What are the advantages when using the second level of "baremetal"?

**Comment by memsharded** [comment]
Well, that is irrelevant now, Conan 2 was released Feb 2023, and it didn't move Arduino to be a nested subsetting of ``os``.

The thing with ``baremetal`` is that for example it will map automatically to a ``"baremetal": "Generic"`` in CMake toolchains (CMAKE_SYSTEM_NAME), while others won't. 


This ticket would need an update for latest Conan 2. There are nowadays some new features, like the possibility to inject arbitrary CMake variables into CMakeToolchain, that could be used for this issue.

**Comment by czoido** [comment]
Hi @serghov,

As mentioned in the last comment, there are several ways to address this in Conan 2 when cross-compiling for bare-metal targets using `CMakeToolchain`. For example, you can configure it directly in the recipe like this:

~~~python
def generate(self):
    ...
    tc = CMakeToolchain(self)
    tc.variables["CMAKE_SYSTEM_NAME"] = "Generic"
    tc.variables["CMAKE_SYSTEM_PROCESSOR"] = "arm"
    tc.variables["CMAKE_TRY_COMPILE_TARGET_TYPE"] = "STATIC_LIBRARY"
    tc.generate()
    ...
~~~

Or alternatively, through the `conf` section in the profile:

~~~ini
...
[conf]
tools.cmake.cmaketoolchain:system_name=Generic
tools.cmake.cmaketoolchain:system_processor=arm
tools.cmake.cmaketoolchain:extra_variables={"CMAKE_TRY_COMPILE_TARGET_TYPE": "STATIC_LIBRARY"}
~~~

Just wondering — is this still an issue for you? Can we provide any more help here?

**Comment by creiter64** [comment]
I just stumbled over the same issue today. Would it be an option to set `"CMAKE_TRY_COMPILE_TARGET_TYPE": "STATIC_LIBRARY"` on the `baremetal` target as default? I'm not if there are drawbacks to this.

**Comment by memsharded** [comment]
Yes, I am afraid that defining ``STATIC_LIBRARY`` as default could be breaking, even for ``baremetal``, because the CMake default (https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html) is ``EXECUTABLE``, so I'd say that there might be indeed many try-compile checks in embedded projects that expect this executable to work, depending on the baremetal platform and linkage approach.

**Comment by memsharded** [comment]
As some extra information, it is likely that using a ``tools.cmake.cmaketoolchain:user_toolchain`` conf is probably one of the recommended ways to do this, as setting a baremetal toolchain is not trivial (depending on the use case, a tools.cmake.cmaketoolchain:toolchain_file`` as commented above could be preferred, I just wanted to introduce the ``user_toolchain`` conf here too)
